File: lena-config/src/test/java/org/riversoforion/lena/config/NamespaceTest.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.NullAndEmptySource;

import java.util.List;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.junit.jupiter.params.provider.Arguments.arguments;

class NamespaceTest {

    @Test
    void root() {

        assertThat(Namespace.root().name()).isEqualTo("/");
    }

    @ParameterizedTest(name = "''{0}''")
    @MethodSource("ofSegments")
    @DisplayName("factory method 'of'")
    void of(String expectedName, String[] segments) {

        assertThat(Namespace.of(segments).name()).isEqualTo(expectedName);
    }

    static Stream<Arguments> ofSegments() {

        return Stream.of(arguments("/first", strings("first")),
                         arguments("/first/second", strings("first", "second")),
                         arguments("/number-1/number-2/number-3", strings("number-1", "number-2", "number-3")),
                         arguments("/two-parts/three-name-parts", strings("two-parts", "three-name-parts")),
                         arguments("/good-segment", strings("good.segment")),
                         arguments("/valid-name", strings("valid_name")),
                         arguments("/uppercaseallowed", strings("UppercaseAllowed")),
                         arguments("/good-segment/also/good/segment", strings("good-segment", "also/good/segment")),
                         arguments("/valid-segment/good-segment", strings("Valid-Segment", "good-segment")),
                         arguments("/this-is-fine/this-one-too/tambien-bueno", strings("this-is-fine", "this-one-too", "también_bueno")),
                         arguments("/spaces-are-allowed/and-numb3rs/and-pec-al-c-aracter-",
                                   strings("spaces are allowed", "and numb3rs", "and $pec!al c#aracter$")),
                         arguments("/", strings()),
                         arguments("/", strings((String) null)),
                         arguments("/", null));
    }

    @ParameterizedTest(name = "''{1}'' -> ''{0}''")
    @CsvSource(textBlock = """
                           /, /
                           /first, /first
                           /first/second, /first/second
                           /number-1/number-2/number-3, /number-1/number-2/number-3
                           /, '      '
                           /, ///
                           /first, first
                           /first/second, first/second
                           /good-segment/valid-segment, /good-segment/ValiD-SegmenT
                           /this-is-fine/this-is-fine, /this_is_fine/this-is-fine
                           /not-wrong, /not.wrong
                           /spaces-and/special-characters/a-w3d, /spaces and/special characters/a!!@w3d
                           /allow/empty-segments, /allow//empty-segments
                           /whitespace/around-segments/is-trimmed, '/ whitespace /\taround segments/is-trimmed '
                           /first/second, ///first//second/
                           """)
    @DisplayName("factory method 'parse'")
    void parse(String expectedName, String name) {

        assertThat(Namespace.parse(name).name()).isEqualTo(expectedName);
    }

    @ParameterizedTest(name = "''{1}'' -> ''{0}''")
    @CsvSource(textBlock = """
                           segment, segment
                           segment-name, segment-name
                           segment-name, Segment-NAME
                           s3gment-name, s3gment name
                           -3gment-nam-, $3gment_nam*
                           segment-name, ' segment name '
                           '', ''
                           '',
                           '', '    '
                           """)
    @DisplayName("normalize segment")
    void normalizeSegment(String expectedResult, String segment) {

        assertThat(Namespace.normalizeSegment(segment)).isEqualTo(expectedResult);
    }

    @ParameterizedTest(name = "''{0}'' resolves to ''{2}''")
    @CsvSource(textBlock = """
                           /parent/sub-1, 2, sub-1
                           /parent/sub-1/sub-2/sub-3, 4, sub-1/sub-2/sub-3
                           /absolute/child, 2, /absolute/child
                           /parent/valid-name, 2, valid name
                           /parent/also-valid, 2, also__valid
                           """)
    @DisplayName("create child namespaces with valid name")
    void child_WithValidName(String expectedChildName, int expectedSegments, String name) {

        Namespace parent = Namespace.of("parent");
        Namespace child = parent.child(name);
        assertThat(child.name()).isEqualTo(expectedChildName);
        assertThat(child.segments()).hasSize(expectedSegments);
    }

    @ParameterizedTest(name = "''{0}''")
    @NullAndEmptySource
    @CsvSource(textBlock = """
                           '    '
                           """)
    @DisplayName("create child namespaces with invalid name")
    void child_WithInvalidName(String name) {

        Namespace parent = Namespace.of("parent");
        assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> parent.child(name));
    }

    @Test
    @DisplayName("property names resolve")
    void resolveProperty() {

        Namespace namespace = Namespace.of("parent", "child");
        List<String> propName = namespace.resolveProperty("prop-name", "yet.anotherProp");
        assertThat(propName).containsExactly("parent", "child", "prop-name", "yet.anotherProp");
    }

    private static Object strings(String... strings) {

        return strings;
    }

    @Test
    @DisplayName("Comparable implementation")
    void comparable() {
        Namespace namespace = Namespace.of("parent", "child");
        Namespace same = Namespace.of("parent", "child");
        Namespace less = Namespace.of("parent", "a");
        Namespace more = Namespace.of("parent", "d");

        assertThat(namespace.compareTo(null)).isPositive();
        assertThat(namespace.compareTo(same)).isZero();
        assertThat(namespace.compareTo(less)).isPositive();
        assertThat(namespace.compareTo(more)).isNegative();
    }

    @Test
    @DisplayName("CharSequence implementation")
    void charSequence() {
        Namespace namespace = Namespace.of("parent", "child");
        assertThat(namespace.length()).isEqualTo(13);
        assertThat(namespace.charAt(0)).isEqualTo('/');
        assertThat(namespace.charAt(10)).isEqualTo('i');
        assertThat(namespace.subSequence(1, 7)).isEqualTo("parent");
    }
}

File: lena-config-api/src/commonTest/kotlin/org/riversoforion/lena/config/NamespaceTest.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class NamespaceTest {

    @Test
    fun root() {
        assertEquals("/", Namespace.root().name())
    }

    @Test
    fun of() {
        val cases = listOf(
            listOf("first") to "/first",
            listOf("first", "second") to "/first/second",
            listOf("number-1", "number-2", "number-3") to "/number-1/number-2/number-3",
            listOf("two-parts", "three-name-parts") to "/two-parts/three-name-parts",
            listOf("good.segment") to "/good-segment",
            listOf("valid_name") to "/valid-name",
            listOf("UppercaseAllowed") to "/uppercaseallowed",
            listOf("good-segment", "also/good/segment") to "/good-segment/also/good/segment",
            listOf("Valid-Segment", "good-segment") to "/valid-segment/good-segment",
            // Accented characters are deliberately excluded here: NFD decomposition (JVM) vs.
            // best-effort stripping (native) produce different results. See platform-specific
            // tests in jvmTest/nativeTest for that behavior.
            listOf("spaces are allowed", "and numb3rs", "and \$pec!al c#aracter\$") to
                "/spaces-are-allowed/and-numb3rs/and-pec-al-c-aracter-",
            emptyList<String>() to "/",
        )
        for ((segments, expected) in cases) {
            assertEquals(expected, Namespace.of(*segments.toTypedArray()).name(), "segments=$segments")
        }
    }

    @Test
    fun parse() {
        val cases = listOf(
            "/" to "/",
            "/first" to "/first",
            "/first/second" to "/first/second",
            "/number-1/number-2/number-3" to "/number-1/number-2/number-3",
            "      " to "/",
            "///" to "/",
            "first" to "/first",
            "first/second" to "/first/second",
            "/good-segment/ValiD-SegmenT" to "/good-segment/valid-segment",
            "/this_is_fine/this-is-fine" to "/this-is-fine/this-is-fine",
            "/not.wrong" to "/not-wrong",
            "/spaces and/special characters/a!!@w3d" to "/spaces-and/special-characters/a-w3d",
            "/allow//empty-segments" to "/allow/empty-segments",
            "/ whitespace /\taround segments/is-trimmed " to "/whitespace/around-segments/is-trimmed",
            "///first//second/" to "/first/second",
        )
        for ((input, expected) in cases) {
            assertEquals(expected, Namespace.parse(input).name(), "input='$input'")
        }
    }

    @Test
    fun child_WithValidName() {
        val parent = Namespace.of("parent")
        val cases = listOf(
            "sub-1" to ("/parent/sub-1" to 2),
            "sub-1/sub-2/sub-3" to ("/parent/sub-1/sub-2/sub-3" to 4),
            "/absolute/child" to ("/absolute/child" to 2),
            "valid name" to ("/parent/valid-name" to 2),
            "also__valid" to ("/parent/also-valid" to 2),
        )
        for ((input, expected) in cases) {
            val (expectedName, expectedSegments) = expected
            val child = parent.child(input)
            assertEquals(expectedName, child.name(), "input='$input'")
            assertEquals(expectedSegments, child.segments().size, "input='$input'")
        }
    }

    @Test
    fun child_WithInvalidName() {
        val parent = Namespace.of("parent")
        for (input in listOf("", "    ")) {
            assertFailsWith<IllegalArgumentException>("input='$input'") { parent.child(input) }
        }
    }

    @Test
    fun resolveProperty() {
        val namespace = Namespace.of("parent", "child")
        val propName = namespace.resolveProperty(Name.of("prop-name", "yet.anotherProp"))
        assertEquals(listOf("parent", "child", "prop-name", "yet.anotherProp"), propName)
    }

    @Test
    fun comparable() {
        val namespace = Namespace.of("parent", "child")
        val same = Namespace.of("parent", "child")
        val less = Namespace.of("parent", "a")
        val more = Namespace.of("parent", "d")

        assertEquals(0, namespace.compareTo(same))
        assertEquals(true, namespace.compareTo(less) > 0)
        assertEquals(true, namespace.compareTo(more) < 0)
    }
}
